File: backend/app/schemas/studio.py
"""
============================================================
AI Product Studio API Schemas
—— POST /api/product/create 等端点的请求/响应契约
============================================================
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class ProductCreateRequest(BaseModel):
    """创建产品（触发 Research → Product → Design → Presentation 流水线）。"""

    idea: str = Field(..., min_length=1, max_length=500, description="产品想法")
    # 模板选择权（前端）：设计主题/风格方法论，空 = LLM 自主决策
    theme_id: str | None = Field(default=None, max_length=64,
                                 description="设计主题 id（THEME_PRESETS）")
    style_id: str | None = Field(default=None, max_length=64,
                                 description="风格方法论 id（ppt-master styles）")


class ProductCreateResponse(BaseModel):
    """创建成功 —— 异步流水线立即返回，前端轮询 GET /api/product/{id}。"""

    product_id: str
    idea: str
    status: str


class ProductAssetResponse(BaseModel):
    """产品资产包查询响应（对齐 POST /api/product/create 的目标响应形状）。"""

    product_id: str
    idea: str
    status: str
    error_message: str | None = None
    created_at: str | None = None
    updated_at: str | None = None
    # ── 六节点结构化资产（未完成时为 None） ──
    requirement: dict[str, Any] | None = None
    research: dict[str, Any] | None = None
    competitor_matrix: dict[str, Any] | None = None
    competitor_analysis: dict[str, Any] | None = None
    strategy: dict[str, Any] | None = None
    design: dict[str, Any] | None = None
    presentation: dict[str, Any] | None = None
    ppt_design: dict[str, Any] | None = None
    document: dict[str, Any] | None = None
    # ── Key Words：任务完成时按「设计/功能/外观/人群/场景」总结，用户可编辑 ──
    keywords: dict[str, list[str]] | None = None
    node_models: dict[str, str] | None = None
    # ── P5: 质量层 ──
    critic_score: int | None = None
    gate_report: dict[str, Any] | None = None
    # ── C5: 各节点模型 token 用量（成本可观测） ──
    usage: dict[str, Any] | None = None
    # ── 进度与失败记录 ──
    node_status: dict[str, str] = Field(default_factory=dict)
    errors: dict[str, str] = Field(default_factory=dict)


class ProductListResponse(BaseModel):
    """产品列表项（轻量）。"""

    product_id: str
    idea: str
    status: str
    created_at: str | None = None
    keywords: dict[str, list[str]] | None = None


class ProductKeywordsUpdateRequest(BaseModel):
    """关键词组编辑请求（整体替换；组键固定为 design/function/appearance/audience/scenario）。"""

    keywords: dict[str, list[str]] = Field(..., description="方面 → 关键词列表")


class ProductKeywordsUpdateResponse(BaseModel):
    """关键词组保存结果。"""

    product_id: str
    keywords: dict[str, list[str]]
    updated: bool = True


class ExportPdfResponse(BaseModel):
    """PPT 风格 PDF 导出结果。"""

    product_id: str
    pdf_url: str
    message: str


class KnowledgeDocumentResponse(BaseModel):
    """知识库文档条目（全局聚合，只读）。"""

    document_id: str
    project_id: str
    project_topic: str
    section_title: str
    section_order: int = 0
    updated_at: str | None = None


class PresentationUpdateRequest(BaseModel):
    """演示编辑器保存：回写 Presentation DSL。"""

    presentation: dict[str, Any] = Field(..., description="完整 Presentation DSL")


class ProductImageSearchRequest(BaseModel):
    """编辑器素材搜索（无状态，DuckDuckGo）。"""

    query: str = Field(..., min_length=1)
    max_results: int = Field(default=12, ge=1, le=20)
    search_depth: int = Field(default=5, ge=5, le=20)


class ProductImageResult(BaseModel):
    """搜索结果条目（不持久化，临时 id）。"""

    id: str
    query: str
    title: str
    image_url: str
    source_url: str | None = None


class ProductImageSearchResponse(BaseModel):
    """编辑器素材搜索结果。"""

    images: list[ProductImageResult]
    total_count: int

File: frontend/src/types/studio.ts
/**
 * ============================================================
 * AI Product Studio 类型定义
 * —— 与 agent-platform 的 Pydantic Schemas 保持同步
 *    （research / product / design / presentation 四类结构化资产）
 * ============================================================
 */

export type StudioStatus =
  | 'queued'
  | 'running'
  | 'paused'
  | 'completed'
  | 'failed'
  | 'waiting_approval'

export type KeywordGroups = Record<string, string[]>

export type SuggestionResponse = {
  suggestions: string[]
}

export type ClarifyRequest = {
  idea: string
  messages: Array<{ role: 'user' | 'assistant'; content: string }>
  max_rounds?: number
}

export interface StudioProductCreateResponse {
  product_id: string
  idea: string
  status: StudioStatus
}

// ─── 市场研究（Research Agent → MarketResearch） ─────────────

export interface MarketSizeInfo {
  summary: string
  tam?: string | null
  sam?: string | null
  som?: string | null
  cagr?: string | null
  source?: string | null
}

export interface CompetitorInfo {
  name: string
  url?: string | null
  positioning?: string | null
}

export interface SourceRef {
  url: string
  title?: string
  weight?: number
}

export interface MarketResearch {
  market_size: MarketSizeInfo
  competitors: CompetitorInfo[]
  customer_pain_points: string[]
  industry_trends: string[]
  sources?: SourceRef[]
}

// ─── 竞品分析（Competitor Analysis → CompetitorAnalysis） ────

export interface CompetitorProfile {
  name: string
  positioning?: string
  target_segment?: string | null
  pricing?: string | null
  strengths: string[]
  weaknesses: string[]
  threat_level: 'high' | 'medium' | 'low'
}

export interface CompetitorMatrixInfo {
  dimensions: string[]
  profiles: CompetitorProfile[]
}

export interface CompetitorAnalysis {
  competitors: CompetitorProfile[]
  matrix: CompetitorMatrixInfo
  competitive_landscape: string
  differentiation_opportunities: string[]
}

// ─── 产品策略（Product Agent → ProductStrategy） ─────────────

export interface Persona {
  name: string
  role?: string
  goals: string[]
  pain_points: string[]
  behavior?: string | null
}

export interface Feature {
  name: string
  description?: string
  category?: string | null
  priority: 'P0' | 'P1' | 'P2'
}

export interface RoadmapItem {
  phase: string
  title: string
  goal?: string | null
  timeline?: string | null
  milestones: string[]
}

export interface PRDSection {
  title: string
  content: string
}

export interface ProductStrategy {
  positioning: string
  personas: Persona[]
  features: Feature[]
  roadmap: RoadmapItem[]
  prd_sections: PRDSection[]
  sources?: SourceRef[]
}

// ─── UX 设计（Design Agent → UXDesign） ──────────────────────

export interface UserFlowStep {
  step: string
  description?: string
  is_entry?: boolean
  is_exit?: boolean
}

export interface PageSpec {
  name: string
  purpose?: string | null
  key_elements: string[]
}

export interface ComponentSpec {
  name: string
  kind?: string
  description?: string | null
}

export interface UXDesign {
  user_flow: UserFlowStep[]
  pages: PageSpec[]
  components: ComponentSpec[]
  sources?: SourceRef[]
}

// ─── 演示（P2/P4: Presentation DSL；旧 SlideDeck 见下方兼容类型） ──

import type {
  PresentationDSL,
  QualityGateReport,
} from '@/types/presentation'

// @deprecated 旧版 SlideDeck（P2 前资产包，兼容展示用）
export type SlideBlockType =
  | 'title' | 'subtitle' | 'text' | 'bullets'
  | 'metric' | 'quote' | 'table' | 'image'

export interface SlideBlock {
  id: string
  block_type: SlideBlockType
  content: string
  emphasis?: 'low' | 'normal' | 'high'
  meta?: Record<string, unknown>
}

export interface Slide {
  id: string
  title: string
  subtitle?: string | null
  layout_type: string
  blocks: SlideBlock[]
  visual_metadata?: Record<string, unknown>
}

export interface DeckSection {
  title: string
  slide_ids: string[]
}

export interface SlideDeck {
  topic: string
  slides: Slide[]
  sections: DeckSection[]
}

// ─── 资产包（GET /api/v1/product/{id} 响应） ─────────────────

export interface StudioProduct {
  product_id: string
  idea: string
  status: StudioStatus
  error_message?: string | null
  created_at?: string | null
  updated_at?: string | null
  keywords?: KeywordGroups | null
  node_status?: Record<string, string>
  node_models?: Record<string, string>
  errors?: Record<string, string>
  requirement?: Record<string, unknown> | null
  research?: MarketResearch | null
  competitor_analysis?: CompetitorAnalysis | null
  strategy?: ProductStrategy | null
  design?: UXDesign | null
  document?: Record<string, unknown> | null
  presentation?: PresentationDSL | SlideDeck | null
  ppt_design?: {
    project_dir?: string
    pptx_path?: string
    pptx_relative?: string
    pages?: number
    created_at?: string
    svg_files?: string[]
    /** 后端 ppt-master 的最终 SVG 缩略图 URL */
    svg_previews?: string[]
    images?: Array<{ name: string; url: string; size: number }>
    model?: string
    design_brief?: string
    status?: string
    /** P7: 资产对账恢复标记 —— 表示此 ppt_design 由磁盘 ppt_projects 恢复而来 */
    recovered?: boolean
  } | null
  critic_score?: number | null
  gate_report?: QualityGateReport | null
}

export interface ExportPdfResponse {
  product_id: string
  pdf_url: string
  message: string
}

/** P7: PPT 资产库索引项（GET /api/v1/product/ppt-assets 响应） */
export interface PptAssetIndexEntry {
  folder_name: string
  title: string
  pptx_url: string
  size: number
  svg_count: number
  created_at?: string | null
  svg_previews: string[]
}

// ─── Design Studio v2（任务级「设计思路 + 图片」资产库） ──────

export type DesignStudioItemKind = 'standalone' | 'component' | 'composite'

export interface DesignStudioImage {
  name: string
  url: string
  size: number
}

export interface DesignStudioVersion {
  ts: string
  text: string
  prompt: string
  image: DesignStudioImage | null
}

export interface DesignStudioItem {
  id: string
  kind: DesignStudioItemKind
  name: string
  /** 设计思路（用户可编辑，重新生图时作为 prompt 主体） */
  text: string
  /** 实际发送给生图模型的完整 prompt（只读） */
  prompt: string
  /** 生图模型返回的文本输出（如 MiniMax data.text，无则 null） */
  api_text: string | null
  image: DesignStudioImage | null
  source: 'pipeline' | 'user'
  parent: string | null
  children: string[]
  created_at: string
  updated_at: string
  versions: DesignStudioVersion[]
}

export interface DesignStudioLibrary {
  schema_version: number
  product_id: string
  idea: string
  status: string
  created_at: string
  updated_at: string
  items: DesignStudioItem[]
}
